File: sdk/golang/sfa/output.go
package sfa

import (
	"encoding/json"
	"fmt"
	"os"
)

// writeResult writes the agent's result to stdout in the specified format.
func writeResult(result any, format OutputFormat) {
	switch format {
	case OutputJSON:
		data, err := json.Marshal(result)
		if err != nil {
			fmt.Fprintf(os.Stderr, "error: failed to marshal result: %v\n", err)
			os.Exit(ExitFailure)
		}
		fmt.Fprintln(os.Stdout, string(data))
	default:
		// Text mode — stringify the result
		switch v := result.(type) {
		case string:
			fmt.Fprintln(os.Stdout, v)
		case map[string]any:
			data, _ := json.MarshalIndent(v, "", "  ")
			fmt.Fprintln(os.Stdout, string(data))
		default:
			data, err := json.Marshal(v)
			if err != nil {
				fmt.Fprintln(os.Stdout, fmt.Sprintf("%v", v))
			} else {
				fmt.Fprintln(os.Stdout, string(data))
			}
		}
	}
}

// writeDiagnostic writes a diagnostic message to stderr.
func writeDiagnostic(message string) {
	fmt.Fprintln(os.Stderr, message)
}

// exitWithError writes an error message to stderr and exits with the given code.
func exitWithError(message string, code int) {
	fmt.Fprintf(os.Stderr, "error: %s\n", message)
	os.Exit(code)
}

// emitProgress writes a progress message to stderr in the SFA format.
func emitProgress(agentName, message string) {
	fmt.Fprintf(os.Stderr, "[agent:%s] %s\n", agentName, message)
}

File: sdk/typescript/@sfa/sdk/output.ts
import type { AgentResult, OutputFormat } from "./types";
import { ExitCode } from "./types";

/**
 * Write the agent result to stdout in the specified format.
 * Diagnostics always go to stderr.
 */
export function writeResult(result: AgentResult, format: OutputFormat): void {
  if (format === "json") {
    const output: Record<string, unknown> = { result: result.result };
    if (result.metadata) output.metadata = result.metadata;
    if (result.warnings && result.warnings.length > 0) output.warnings = result.warnings;
    if (result.error) output.error = result.error;
    process.stdout.write(JSON.stringify(output) + "\n");
  } else {
    // Text mode: write result as string
    const text = typeof result.result === "string" ? result.result : JSON.stringify(result.result, null, 2);
    process.stdout.write(text + "\n");
  }
}

/**
 * Write a diagnostic/error message to stderr.
 */
export function writeDiagnostic(message: string): void {
  process.stderr.write(message + "\n");
}

/**
 * Write an error and exit with the appropriate code.
 */
export function exitWithError(message: string, code: number = ExitCode.FAILURE): never {
  writeDiagnostic(`error: ${message}`);
  process.exit(code);
}

/**
 * Emit a progress message to stderr in the standard format.
 */
export function emitProgress(agentName: string, message: string): void {
  process.stderr.write(`[agent:${agentName}] ${message}\n`);
}
